python/ray/data/dataset.py

```diff
@@ -746,16 +748,26 @@ def add_column(
                 ray (e.g., num_gpus=1 to request GPUs for the map tasks).
         """
 
-        def add_column(batch: "pandas.DataFrame") -> "pandas.DataFrame":
-            batch.loc[:, col] = fn(batch)
-            return batch
+        def add_column(batch: "pyarrow.Table") -> "pyarrow.Table":
+            # fn can output either a pandas Series or a pyarrow Array
+            column = fn(batch)
+
+            # The index of the column will be -1 if it is missing in which case we'll
+            # want to append it
+            column_idx = batch.schema.get_field_index(col)
+            if column_idx == -1:
+                # Append the column to the table
+                return batch.append_column(col, column)
+            else:
+                # Create a new table with the updated column
+                return batch.set_column(column_idx, col, column)
 
         if not callable(fn):
             raise ValueError("`fn` must be callable, got {}".format(fn))
 
         return self.map_batches(
             add_column,
-            batch_format="pandas",  # TODO(ekl) we should make this configurable.
+            batch_format="pyarrow",
```

Comment:
As discussed offline, we are not making this configurable as it's an implementation detail.

Reply:
Ah, misunderstood in our earlier discussion. If this is specific to `add_column`, it might make sense to expose the batch format since the user is providing a UDF.

Reply:
This is sort of an aside, but IMO `add_column` doesn't provide much value over `map_batches`

Reply:
Ok. Before we only accepted Dataframes so I assume the following makes sense:
We should default to pyarrow.
We should also accept pandas.
We should not accept other data formats.